From 8ffd854411921ab99e9b92badbf4e2c050b44163 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:00:44 -0400
Subject: [ethernet] mlx4: Load the Eth driver first

Message-id: <16dc6509af320afff1c9112dc1e16df01093b2e6.1410425016.git.avadai@redhat.com>
Patchwork-id: 93662
O-Subject: [Patch RHEL7.1 V1 108/197] net/mlx4_core: Load the Eth driver first
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Or Gerlitz <ogerlitz@mellanox.com>

When running in SRIOV mode, VM that is assigned with a non-provisioned
Ethernet VFs get themselves a random mac when the Eth driver starts. In
this case, if the IB driver startup code that deals with RoCE runs first,
it will use a zero mac as the source mac for the Para-Virtual CM MADs
which is buggy. To handle that, we change the order of loading.

Signed-off-by: Or Gerlitz <ogerlitz@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit f24f790f8eb8bca00c66781b21de2a9ff7cd1c00)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/main.c b/drivers/net/ethernet/mellanox/mlx4/main.c
index bbd9c3c..8200cd0 100644
--- a/drivers/net/ethernet/mellanox/mlx4/main.c
+++ b/drivers/net/ethernet/mellanox/mlx4/main.c
@@ -754,10 +754,10 @@ static void mlx4_request_modules(struct mlx4_dev *dev)
    has_eth_port = true;
  }
 
- if (has_ib_port || (dev->caps.flags & MLX4_DEV_CAP_FLAG_IBOE))
-  request_module_nowait(IB_DRV_NAME);
  if (has_eth_port)
   request_module_nowait(EN_DRV_NAME);
+ if (has_ib_port || (dev->caps.flags & MLX4_DEV_CAP_FLAG_IBOE))
+  request_module_nowait(IB_DRV_NAME);
 }
 
 /*
-- 
1.7.1